sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -1164,4 +1167,23 @@ object DateTimeUtils {
     threadLocalTimestampFormat.remove()
     threadLocalDateFormat.remove()
   }
+
+  class MicrosCalendar(tz: TimeZone) extends GregorianCalendar(tz, Locale.US) {
+    def getMicros(digitsInFraction: Int): SQLTimestamp = {
```

Comment:
IIUC, here we are trying to simulate the Java 8 DateTimeFormatter behavior, although we use FastDateFormat under the hood.

what's the behavior of java 8 DateTimeFormatter if the actual value has more digits in the fraction than the number of `S` in the format string?

Reply:
It depends on the mode: https://github.com/apache/spark/blob/60be6d2ea3560143515c1ce9d0a7da416f8f595a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala#L80 . We set `strict` in the master, so, Spark should rejects such input.

Reply:
from the doc:
```
Fraction: Outputs the nano-of-second field as a fraction-of-second. The nano-of-second value has nine digits, thus the count of pattern letters is from 1 to 9. If it is less than 9, then the nano-of-second value is truncated, with only the most significant digits being output.
```

So we need to truncate.

Can we add some comments to explain what we are doing here? It's really tricky.

Reply:
another question is the leading 0s. As a fraction, `1.2` and `1.20` has no difference. Is it also true to timestamp seconds?

Reply:
Correct, I will add special checks for this:
```
    check("yyyy-MM-dd'T'HH:mm:ss.SX", "2019-10-14T09:39:07.1Z", "2019-10-14T09:39:07.1Z")
    check("yyyy-MM-dd'T'HH:mm:ss.SSX", "2019-10-14T09:39:07.10Z", "2019-10-14T09:39:07.1Z")
```